### PR TITLE
#164139768 Send verification email 164139768

### PR DESCRIPTION
**Pull request by devskope**
**What does this PR do?**
Sends verification emails to users after successful registration

**Description of Task to be completed?**
- Send mail to signed up user
-  add endpoint to verify User accounts

**How should this be manually tested?**
- Checkout `ft-send-verification-email-164139768`
- Run `npm i`
- Create a new user account by sending a valid post request to `/api/users`
  A successfully created user should receive a verification email.

**Any background context you want to provide?**
N/A

**What are the relevant pivotal tracker stories?**
#164139768

**Screenshots (if appropriate)**

### CODE REVIEW

File: server/models/User.js [hunk 1]
```diff
@@ -11,35 +11,36 @@ module.exports = (sequelize, DataTypes) => {
...
-        validate: { isUrl: { msg: '"image" must be a valid URL' } }
-      }
+        validate: { isUrl: { msg: '"image" must be a valid URL' } },
+      },
     },
     {}
   );
   User.beforeValidate(user => {
-    user.hash = hashPassword(user.hash);
+    if (user.hash) user.hash = hashPassword(user.hash);
```
Comment (LordUche) [inline server/models/User.js]: Why did you add an `if` statement here?

File: server/utils/mailer.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * @returns {Promise} Sendgrid response
+ */
+
+const sendVerificationMail = ({ email, username, verificationLink }) => {
+  const mail = {
+    to: email,
+    from: ADMIN_EMAIL,
+    text: `Please follow this link to activate your ${siteName} account:   ${verificationLink}`,
+    html: ' ',
+    templateId: 'd-6e841534b95240fa929b2fcd78a8ff7f',
```
Comment (LordUche) [inline server/utils/mailer.js:23]: Where did you get this `templateId`? Is it a random string?

File: test/users.spec.js [hunk 3]
```diff
@@ -117,4 +121,29 @@ describe('Authentication', () => {
...
+        });
+    });
+
+    it('should check that the token is valid', done => {
+      chai
+        .request(server)
+        .get(`${baseUrl}?token=io.fjf.tir`)
+        .end((err, res) => {
+          expect(res).to.have.status(500);
+          expect(res.body.verified).to.equal(undefined);
```
Comment (LordUche) [inline test/users.spec.js]: Why `undefined` and not `false`?

**Review by LordUche — DISMISSED** [review]

File: package.json [hunk 4]
```diff
@@ -52,16 +53,16 @@
     "coveralls": "3.0.3",
     "eslint": "5.15.0",
     "eslint-config-airbnb-base": "13.1.0",
-    "eslint-config-prettier": "^4.1.0",
+    "eslint-config-prettier": "4.1.0",
```
Comment (andela-pessien) [inline package.json:57]: When did we decide to use prettier? 🙂

File: server/models/User.js (hunk 1 above)
Comment (andela-pessien) [reply server/models/User.js]: I think the aim is to avoid rehashing an already hashed password. I would advise to confirm that `user.hash` only exists here when the field is created/updated.

File: server/utils/mailer.js (hunk 2 above)
Comment (andela-pessien) [reply server/utils/mailer.js:23]: I'm assuming from context that this is a Sendgrid template - that should be made clear (no magic values please 🙂 ). Preferably would include a link to the template as well.

I like the initiative of using a template hosted online though - that allows you to make style and other changes to the email without having to redeploy your app.

File: test/users.spec.js (hunk 3 above)
Comment (andela-pessien) [reply test/users.spec.js]: Seconding this

File: server/config/env-config.js [hunk 5]
```diff
@@ -2,11 +2,32 @@ import dotenv from 'dotenv';
...
+  AUTH_TOKEN_EXPIRY,
+  VERIFICATION_LINK_EXPIRY,
+} = env;
 
 export default {
   DATABASE_URL,
   NODE_ENV,
   PORT,
-  SECRET
+  SECRET,
```
Comment (andela-pessien) [inline server/config/env-config.js:24]: This file is getting a bit repetitive, no?

File: server/controllers/auth-controller.js [hunk 6]
```diff
@@ -33,13 +45,31 @@ class UsersController {
         try {
           const { dataValues: user } = await User.create({
             ...body,
-            hash: body.password
+            hash: body.password,
           });
           const { id, username } = user;
-          const token = generateToken({ id, username });
+          const [token, verificationToken] = await authTokens({
```
Comment (andela-pessien) [inline server/controllers/auth-controller.js]: What's the purpose of the two tokens here?

File: server/utils/helpers.js [hunk 7]
```diff
@@ -15,6 +20,24 @@ const { SECRET } = env;
...
+/**
+ * Generates account verification and auth tokens
+ *
+ * @param {object} options - token generation options.
+ * @param {string} options.payload - payload to sign
+ * @param {string} options.tokenExpiresIn - auth token expiry
+ * @param {string} options.verificationTokenExpiresIn - verification token expiry
+ * @returns {Promise.<[string,string]>} [authToken, verificationToken]
+ */
+export const authTokens = async ({
```
Comment (andela-pessien) [inline server/utils/helpers.js]: Why a single function that returns both?

File: test/users.spec.js [hunk 8]
```diff
@@ -117,4 +121,29 @@ describe('Authentication', () => {
...
+        .request(server)
+        .get(`${baseUrl}?token=${generateToken(tokenPayload)}`)
+        .end((err, res) => {
+          expect(res).to.have.status(200);
+          expect(res.body.verified).to.equal(true);
+          done(err);
+        });
+    });
+
+    it('should check that the token is valid', done => {
```
Comment (andela-pessien) [inline test/users.spec.js]: Would also like to see a test for when there's no token passed

**Review by andela-pessien — CHANGES_REQUESTED** [review]

File: test/users.spec.js (hunk 3 above)
Comment (LordUche) [reply test/users.spec.js]: Is the user not expecting an error message?

File: server/utils/helpers.js (hunk 7 above)
Comment (devskope) [reply server/utils/helpers.js]: Convenience,  seeing as both are needed in the same controller.

File: server/config/env-config.js (hunk 5 above)
Comment (devskope) [reply server/config/env-config.js:24]: This will be addressed as soon as possible.
We'll probably do a conditional export of vars based on current `NODE_ENV`

File: test/users.spec.js (hunk 3 above)
Comment (devskope) [reply test/users.spec.js]: @blackdevelopa  ☝️ 
@LordUche  It is `undefined` because the `verified` property gets tacked on to the response body only when verification succeeds.
I also agree that we need to approach testing more verbosely......  if it's being sent out in the response, test it.

File: server/utils/mailer.js (hunk 2 above)
Comment (devskope) [reply server/utils/mailer.js:23]: @andela-pessien Noted.  We'll share links in internal comms.
@LordUche  Not random string, SendGrid template ID

File: test/users.spec.js (hunk 3 above)
Comment (andela-pessien) [reply test/users.spec.js]: Why would the `verified` property ever be `undefined`? If verification fails then the answer to whether or not the link was `verified` is `false`, no?

File: server/utils/helpers.js (hunk 7 above)
Comment (andela-pessien) [reply server/utils/helpers.js]: I'm not really sure how it is a convenience to generate _both_ with a single function call that requires arguments to be passed for both even though they're not both being used because auth and verification tokens are used in the same file, actually.

File: test/users.spec.js (hunk 8 above)
Comment (andela-pessien) [reply test/users.spec.js]: Has this been implemented?

File: server/config/env-config.js (hunk 5 above)
Comment (andela-pessien) [reply server/config/env-config.js:24]: Please do so - I'm pretty sure the team can come up with a more efficient way than writing out the same list of variables twice. Not required for this PR, but I don't expect to still see this by the end of the week.

Also the existence of one environment variable shouldn't influence the existence of another.

File: server/controllers/auth-controller.js (hunk 6 above)
Comment (devskope) [reply server/controllers/auth-controller.js]: one has a shorter lifespan than the other..... they both are also configurable via external config vars

Comment (andela-pessien) [reply server/controllers/auth-controller.js]: I am quite aware of the difference between them. I'm asking what the purpose of the (authentication) token is here.

File: server/utils/helpers.js (hunk 7 above)
Comment (devskope) [reply server/utils/helpers.js]: The arguments passed for both are completely optional.  they(arguments) only exist to allow for dynamic configuration of token lifespan without app redeployment.

Comment (andela-pessien) [reply server/utils/helpers.js]: So if someone wants to create a verification token with a custom lifespan, they need to call e.g `authTokens(payload, undefined, '1h')`, which returns an array which they then destructure to take the second element as the token they want and discard the first element as they didn't actually want it in the first place but it got generated anyway.

I'm still not really sure how this is for convenience. What exactly is the use case for generating an authentication token and a verification token at the same time?

File: test/users.spec.js (hunk 8 above)
Comment (devskope) [reply test/users.spec.js]: Yes it has been implemented

File: server/controllers/auth-controller.js (hunk 6 above)
Comment (devskope) [reply server/controllers/auth-controller.js]: The authentication token would be required to validate user access to protected routes.

File: test/users.spec.js (hunk 3 above)
Comment (devskope) [reply test/users.spec.js]: The latest commit updates the test and addresses this

File: package.json (hunk 4 above)
Comment (devskope) [reply package.json:57]: 😶 @LordUche 🏃

File: server/utils/helpers.js (hunk 7 above)
Comment (devskope) [reply server/utils/helpers.js]: This function has been dropped and the generation of the tokens has been decoupled in the latest commit

**Review by andela-pessien — APPROVED** [review]
